File: src/main/swift/g0101_0200/s0189_rotate_array/Solution.swift
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Array #Math #Two_Pointers
// #Algorithm_I_Day_2_Two_Pointers #Udemy_Arrays #Big_O_Time_O(n)_Space_O(1)
// #2024_06_28_Time_118_ms_(91.36%)_Space_22.9_MB_(98.87%)

class Solution {
    func rotate(_ nums: inout [Int], _ k: Int) {
        guard nums.count > 1, k > 0, k != nums.count else {
            return
        }

        let requiredRotationsCount = k % nums.count
        let pivotIndex = nums.count - requiredRotationsCount

        reverse(&nums, startIndex: nums.startIndex, endIndex: nums.endIndex - 1)
        reverse(&nums, startIndex: nums.startIndex, endIndex: requiredRotationsCount - 1)
        reverse(&nums, startIndex: requiredRotationsCount, endIndex: nums.endIndex - 1)

    }

    func reverse(_ nums: inout [Int], startIndex: Int, endIndex: Int) {
        var startIndex = startIndex
        var endIndex = endIndex

        while endIndex > startIndex {
            let temp = nums[startIndex]
            nums[startIndex] = nums[endIndex]
            nums[endIndex] = temp
            startIndex += 1
            endIndex -= 1
        }
    }
}

File: src/main/dart/g0101_0200/s0189_rotate_array/Solution.dart
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Array #Math #Two_Pointers
// #Algorithm_I_Day_2_Two_Pointers #Udemy_Arrays #Big_O_Time_O(n)_Space_O(1)
// #2024_10_08_Time_343_ms_(93.58%)_Space_160.6_MB_(19.27%)

class Solution {
  void reverse(List<int> nums, int l, int r) {
    while (l <= r) {
      int temp = nums[l];
      nums[l] = nums[r];
      nums[r] = temp;
      l++;
      r--;
    }
  }

  void rotate(List<int> nums, int k) {
    int n = nums.length;
    int t = n - (k % n);
    reverse(nums, 0, t - 1);
    reverse(nums, t, n - 1);
    reverse(nums, 0, n - 1);
  }
}
